src/rsa/padding.rs

```diff
@@ -159,3 +165,316 @@ pkcs1_digestinfo_prefix!(
+
+        debug_assert!(em_bits >= (8 * h_len) + (8 * s_len) + 9);
+
+        Ok(PSSMetrics {
+            em_len: em_len,
+            db_len: db_len,
+            ps_len: ps_len,
+            s_len: s_len,
+            h_len: h_len,
+            top_byte_mask: top_byte_mask,
+        })
+    }
+}
+
+// Mask-generating function MGF1 as described in
+// https://tools.ietf.org/html/rfc3447#appendix-B.2.1.
+fn mgf1(digest_alg: &'static digest::Algorithm, seed: &[u8], mask: &mut [u8])
+        -> Result<(), error::Unspecified> {
+    let digest_len = digest_alg.output_len;
+
+    // Maximum counter value is the value of (mask_len / digest_len) rounded up.
+    let ctr_max = (mask.len() - 1) / digest_len;
+    assert!(ctr_max <= u32::max_value() as usize);
+    for (i, mask_chunk) in mask.chunks_mut(digest_len).enumerate() {
+        let mut ctx = digest::Context::new(digest_alg);
+        ctx.update(seed);
+        ctx.update(&polyfill::slice::be_u8_from_u32(i as u32));
+        let digest = ctx.finish();
+        let mask_chunk_len = mask_chunk.len();
+        mask_chunk.copy_from_slice(&digest.as_ref()[..mask_chunk_len]);
+    }
+
+    Ok(())
+}
+
+fn pss_digest(digest_alg: &'static digest::Algorithm, msg: &[u8], salt: &[u8])
+              -> digest::Digest {
+    // Fixed prefix.
+    const PREFIX_ZEROS: [u8; 8] = [0u8; 8];
+
+    // Steps 1 & 2 for both encoding and verification. Step 1 is delegated to
+    // the digest implementation.
+    let m_hash = digest::digest(digest_alg, msg);
+
+    // Encoding step 5 and 6, Verification step 12 and 13.
+    let mut ctx = digest::Context::new(digest_alg);
+    ctx.update(&PREFIX_ZEROS);
+    ctx.update(m_hash.as_ref());
+    ctx.update(salt);
+    ctx.finish()
+}
+
+macro_rules! rsa_pss_padding {
+    ( $PADDING_ALGORITHM:ident, $digest_alg:expr, $doc_str:expr ) => {
+        #[doc=$doc_str]
+        /// Feature: `rsa_signing`.
+        pub static $PADDING_ALGORITHM: PSS = PSS {
+            digest_alg: $digest_alg,
+        };
+    }
+}
+
+rsa_pss_padding!(RSA_PSS_SHA256, &digest::SHA256,
+                 "PSS padding using SHA-256 for RSA signatures.");
+rsa_pss_padding!(RSA_PSS_SHA384, &digest::SHA384,
+                 "PSS padding using SHA-384 for RSA signatures.");
+rsa_pss_padding!(RSA_PSS_SHA512, &digest::SHA512,
+                 "PSS padding using SHA-512 for RSA signatures.");
+
+#[cfg(test)]
+mod test {
+    use {error, test};
+    use super::*;
+    use untrusted;
+
+    // Tests PSS verification for variable public modulus lengths
+    #[test]
+    fn test_pss_verify() {
+
+        test::from_file("src/rsa/pss_verify_tests.txt",
+                        |section, test_case| {
+            assert_eq!(section, "");
+
+            let digest_name = test_case.consume_string("Digest");
+            let alg = match digest_name.as_ref() {
+                "SHA256" => &RSA_PSS_SHA256,
+                _ =>  { panic!("Unsupported digest: {}", digest_name) }
```

Comment:
I only wrote these patches to understand the code better and as penance for taking so long to review the code carefully. I'd be very happy if you would take these changes and build more improvements on top of them.

In particular, we do need to add (1) the negative test vectors, (2) padding tests for encoding, hopefully using the same test vectors as are used for padding verification, and (3) end-to-end tests for verification (in rsa_pss_verify_tests.txt) and encoding (in rsa_pss_sign_tests.txt) for non-even modulus lengths, particular {2048, 4096, 8192} ± {1, 7} bits.